src/result.ts

```diff
@@ -2,48 +2,61 @@ import { Maybe, nullable, Nothing } from './maybe'
 
 export type Result<O, E> = {
   ap: <A>(r: Result<A, E>) => Result<A, E>
-  map: <A>(cb: (arg?: O) => A) => Result<A, E>
-  mapErr: <A>(cb: (arg?: E) => A) => Result<O, A>
-  chain: <A>(cb: (arg?: O) => Result<A, E>) => Result<A, E>
-  chainErr: <A>(cb: (arg?: E) => Result<O, A>) => Result<O, A>
+  map: <A>(cb: (arg: O) => A) => Result<A, E>
+  mapErr: <A>(cb: (arg: E) => A) => Result<O, A>
+  chain: <A>(cb: (arg: O) => Result<A, E>) => Result<A, E>
+  chainErr: <A>(cb: (arg: E) => Result<O, A>) => Result<O, A>
   swap: () => Result<O, E>
-  bimap: <A, B>(ok: (arg?: O) => A, err: (arg?: E) => B) => Result<A, B>
+  bimap: <A, B>(ok: (arg: O) => A, err: (arg: E) => B) => Result<A, B>
   cata: <A, B>(obj: {
-    Ok: (arg?: O) => A
-    Err: (arg?: E) => B
-  }) => A|B
+    Ok: (arg: O) => A
+    Err: (arg: E) => B
+  }) => A | B
   toMaybe: () => Maybe<O>
   inspect: () => string
   isErr: () => boolean
   isOk: () => boolean
 }
 
-export const Ok = <O>(arg?: O): Result<O, any> => ({
+
+const innerOk = <O>(arg: O): Result<O, any> => ({
   ap: <A>(r: Result<A, any>) => typeof arg === 'function' ? r.map(x => arg(x)) : Err(),
-  map: <A>(cb: (a?: O) => A): Result<A, any> => Ok(cb(arg)),
-  mapErr: (): Result<O, any> => Ok(arg),
-  chain: <A>(cb: (a?: O) => Result<A, any>): Result<A, any> => cb(arg),
-  chainErr: () => Ok(arg),
-  swap: () => Err(arg),
-  bimap: (ok, _) => Ok(ok(arg)),
+  map: <A>(cb: (a: O) => A): Result<A, any> => innerOk(cb(arg)),
+  mapErr: (): Result<O, any> => innerOk(arg),
+  chain: <A>(cb: (a: O) => Result<A, any>): Result<A, any> => cb(arg),
+  chainErr: () => innerOk(arg),
+  swap: () => innerErr(arg),
+  bimap: (ok, _) => innerOk(ok(arg)),
   cata: obj => obj.Ok(arg),
   toMaybe: () => nullable(arg),
   inspect: () => `Ok(${arg})`,
   isErr: () => false,
   isOk: () => true
 })
 
-export const Err = <E>(arg?: E): Result<any, E> => ({
-  ap: () => Err(arg),
-  map: () => Err(arg),
-  mapErr: <A>(cb: (a?: E) => A): Result<any, A> => Err(cb(arg)),
-  chain: () => Err(arg),
+const innerErr = <E>(arg: E): Result<any, E> => ({
+  ap: () => innerErr(arg),
+  map: () => innerErr(arg),
+  mapErr: <A>(cb: (a: E) => A): Result<any, A> => innerErr(cb(arg)),
+  chain: () => innerErr(arg),
   chainErr: cb => cb(arg),
-  swap: () => Ok(arg),
-  bimap: (_, err) => Err(err(arg)),
+  swap: () => innerOk(arg),
+  bimap: (_, err) => innerErr(err(arg)),
   cata: obj => obj.Err(arg),
   toMaybe: () => Nothing,
   inspect: () => `Err(${arg})`,
   isErr: () => true,
   isOk: () => false
-})
\ No newline at end of file
+})
+
+export function Ok(): Result<any, any>
+export function Ok<O>(arg: O): Result<O, any>
+export function Ok<O>(arg?: O): Result<O | undefined, any> {
+  return innerOk(arg)
+}
+
+export function Err(): Result<any, any>
+export function Err<E>(arg: E): Result<any, E>
+export function Err<E>(arg?: E): Result<any, E | undefined> {
+  return innerErr(arg)
+}
```

Comment:
When running the production build, there are only 2 defined for each Ok and Err, which makes me wonder if we need to define 3 here? Maybe only defining 2 can cover all the cases. Can you confirm this by testing all three cases with the `yarn build` output?

![Screen Shot 2022-10-24 at 9 54 14 AM](https://user-images.githubusercontent.com/9722062/197542688-3952f814-4e0e-41d8-8ae0-61f6fd8f84e7.png)

Reply:
I did some poking around. At runtime, of course, only the one `Ok(arg)` function is needed (and emitted by the compiler as JS), because calling `Ok()` and `Ok(undefined)` are equivalent. Given that, I'd guess that the other two declarations in the typings file are sufficient to reconstruct the type signature in the source TS. I can't eliminate any of the three source function signatures without compiler errors.